### ISSUE TITLE
Grid visualizers for 2D data

### ISSUE THREAD

**Issue by int19h** [Microsoft/PTVS#747]
There should be a facility to conveniently visualize a variable or expression that represents some form of 2D data (numpy arrays, pandas dataframes, tables, matrices etc) in any context where variables and expressions can be inspected and evaluated - i.e. debugging and REPL. 

Some specifics:
- Excel-like grid with row & column headers, selection, scrolling, selection, sorting, filtering etc.
- One tool window per variable/expression, non-modal, dockable.
- Easy to create directly from Locals/Watch/Autos/... (ideally a command in the context menu for entries there).
- Inline editing of elements
- Scales to handle large data sets (think 10000x10000 arrays, and tables of millions of rows). So completely virtualized on both axes, dynamically loading data as needed.
- Heat maps

Some UI mockups:

![image](https://cloud.githubusercontent.com/assets/1175142/9265049/aa247be6-41e2-11e5-85c4-d53e096566d2.png)

![image](https://cloud.githubusercontent.com/assets/1175142/9265046/9d893598-41e2-11e5-8a82-58f161412e93.png)

![image](https://cloud.githubusercontent.com/assets/1175142/9265055/ba9d8652-41e2-11e5-8afe-c4a55e966bb5.png)

**Comment by int19h** [comment]
More detailed spec will follow shortly.

**Comment by zooba** [comment]
IMHO, the heatmap is not a strictly necessary feature, and is actually considerably more complicated than may be expected. There are a ridiculously huge number of accessibility issues to be aware of, as well as many customization points.

I know I'm preempting the detailed spec, but I think there's also a lot we can unify between this and #748 as far as providing frame for an expression and letting the backend do the rendering:

![vizmockup](https://cloud.githubusercontent.com/assets/1693688/9279112/09dc52fc-426a-11e5-9ec0-22feb687677b.png)

The UIElement can also handle whatever virtualization needs to be done, and doesn't have to be baked into the core window.

**Comment by zooba** [comment]
Also be aware that the standard WPF grid view is basically impossible to theme correctly. Azure team didn't even try (table storage viewer, for example), but we really ought to do better than them. On the plus side, even by redefining the control template ourselves we'll be able to better optimize performance.

**Comment by int19h** [comment]
How much of the UX can be unified is an open question. One particular point is that this one needs a single expression (and visualizes its value), while plots need a bunch of statements (and visualize the plotting state machine change that they produce). So I'm not sure whether the editing experience, for example, can truly be combined. 

The standard WPF grid is basically unusable for this purpose. They do scale reasonably well with rows, but they are completely unprepared for even a couple hundred of columns, much less thousands (and I'm not even sure they are properly virtualizing them). It gets even worse when you start selecting things.  And their data binding scheme for cells is also row-centric, which is not a showstopper but is annoying enough.

Either way, just based on my experience with perf, we'll need our own grid widget, and based on what I've read from other people implementing that, it has to avoid the use of WPF element trees for the grid itself if we want anything approaching reasonable perf.

**Comment by int19h** [comment]
The reason why heat map is there is mainly because Spyder has it. I don't think it absolutely needs to be there for v1, but it looks simple enough to implement on the basic level that is available in Spyder. Ideally this is something I'd like user feedback on wrt how useful it is.

**Comment by zooba** [comment]
I think we can get away with WPF controls if we virtualize scrolling properly and have fixed controls for the current visible cells. Drag-select-scrolling is always going to be a challenge, but I don't think saying "only shift+click works to select a range" will be a showstopper.

What if the Expression field could be expanded (like in Excel) to multiple lines? So you see one line normally, but can expand it to view the whole thing. Also a good thing to get feedback on.

**Comment by int19h** [comment]
I believe that is the approach that the WPF datagrid uses. It doesn't scale all that well when someone maximizes the grid on a large screen - you can easily end up with a few hundred cells that way. 

The problem with selection is not so much performing it, as the effect that it has. The way the stock grid handles it, scrolling gets progressively slower the larger the area that you select (and selection itself also gets noticeable delays). 

Expanding the Expression field is doable, but I suspect that we would also want it always expanded for plots (because it usually takes more that one line to set one up). And because those lines are often relatively short, having it at the top like that might not look good, which is why my prototype had it on the left side instead. But the button could also move it to the side when expanded, effectively giving two layouts to work with. I can see advantages in either, and perhaps for complex graphs we really also just need to let people switch back and forth a la XAML designer, because having code and viz side by side will look too crammed.

**Comment by zooba** [comment]
Yeah, that's not difficult to achieve. Two toggle buttons at the right of the text box for left/top and tall/short or wide/narrow. I can see some charts (time series) preferring to keep the expression at the top.

**Comment by syagev** [comment]
In response to @zooba [commenting](https://github.com/Microsoft/PTVS/issues/81#issuecomment-335197399) on #81, I think using matplotlib's popup window or OpenCV's `imshow` to be able to view images stored in numpy arrays is a must for every computer vision / deep-learning developer. 

If there is room to contribute on this issue I'm happy to give it a try but will need a bit pointing to get started...

**Comment by johndpope** [comment]
finally some progress - https://github.com/agermanidis/livepython/issues/5

**Comment by johndpope** [comment]
fyi - @Ladvien made a dataframe preview for r using microsoft excel viewer for vscode.

https://github.com/Ikuyadeu/vscode-R/blob/master/images/DataframePreview.gif

https://github.com/Ikuyadeu/vscode-R/issues/15
